### PR TITLE
Submission proxy as wrapper of Distributor and LogListRefresher.

### PR DESCRIPTION
**Pull request by Mercurrent**
Let's discuss testing approaches.

### CODE REVIEW

File: submission/proxy.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+func (p *Proxy) RestartDistributor(ctx context.Context, ll *loglist.LogList) error {
+	d, err := getDistributor(p.ctPlc)(ll)
+	if err != nil {
+		// losing ll info.
+		return err
+	}
+
+	// Start refreshing roots periodically so they stay up-to-date.
+	go Every(ctx, p.rootsRefreshInterval, func(ctx context.Context) {
+		if errs := d.RefreshRoots(ctx); len(errs) > 0 {
```
Comment (RJPercival) [inline submission/proxy.go]: If `d` was created by an fake `Distributor` factory that was provided by your test, then you could do something like this for testing;
```go
type fakeDistributor struct {
  RefreshRootsWg sync.WaitGroup
}
func (d *fakeDistributor) RefreshRoots(ctx context.Context) ([]error) {
  d.RefreshRootsWg.Done()
}

func TestRestartDistributor(t *testing.T) {
  fake := fakeDistributor{}
  fake.RefreshRootsWg.Add(1)
  ...
  p.RestartDistributor(...)
  ...
  fake.RefreshRootsWg.Wait()
}
```

File: submission/proxy.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+// NewProxy creates and inits a Proxy instance.
+func NewProxy(ctx context.Context, llPath string, llRefresh time.Duration, rootsRefresh time.Duration, ctPlc CTPolicyType) *Proxy {
+	var p Proxy
+	p.logListPath = llPath
+	p.llRefreshInterval = llRefresh
+	p.rootsRefreshInterval = rootsRefresh
+	p.ctPlc = ctPlc
+	p.llWatcher = NewLogListRefresher(llPath)
+
+	go Every(ctx, p.llRefreshInterval, func(ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (Mercurrent) [reply submission/proxy.go]: Added method Run()

File: submission/proxy.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	ctPlc                CTPolicyType
+
+	llWatcher *LogListRefresher
+
+	mu   sync.RWMutex // guards the distributor
+	dist *Distributor
+}
+
+// NewProxy creates and inits a Proxy instance.
+func NewProxy(ctx context.Context, llPath string, llRefresh time.Duration, rootsRefresh time.Duration, ctPlc CTPolicyType) *Proxy {
```
[the remark this answers is not in the document]
Comment (Mercurrent) [reply submission/proxy.go]: Done.

File: submission/proxy.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+	go Every(ctx, p.llRefreshInterval, func(ctx context.Context) {
+		p.RefreshLogList(ctx)
+	})
+
+	return &p
+}
+
+// RefreshLogList reads Log List one time and runs updates if necessary.
+func (p *Proxy) RefreshLogList(ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (Mercurrent) [reply submission/proxy.go]: Added error to the return section

File: submission/proxy.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	if ll, err := p.llWatcher.Refresh(); err != nil {
+		glog.Error(err)
+	} else {
+		p.RestartDistributor(ctx, ll)
+	}
+}
+
+// RestartDistributor activates new Distributor instance with Log List provided
+// and sets it as active.
+func (p *Proxy) RestartDistributor(ctx context.Context, ll *loglist.LogList) error {
```
[the remark this answers is not in the document]
Comment (Mercurrent) [reply submission/proxy.go]: Removed from export.

**Comment by Mercurrent** [comment]
Added some tests (not complete). WDYT on the current approach?

**Comment by Mercurrent** [comment]
friendly ping

**Comment by Mercurrent** [comment]
Passing to Pierre as Rob is OOO

**Comment by Mercurrent** [comment]
ping

**Comment by Mercurrent** [comment]
pong

**Pull request merged**